eth2near/eth2near-block-relay-rs/src/eth2near_relay.rs

```diff
@@ -140,60 +149,130 @@ impl Eth2NearRelay {
         let mut start_slot = start_slot;
         let mut last_slot = last_slot;
         while start_slot + 1 < last_slot {
             let mid_slot = start_slot + (last_slot - start_slot) / 2;
-            if self.block_known_on_near(mid_slot)? {
-                start_slot = mid_slot;
-            } else {
-                last_slot = mid_slot;
+            match self.block_known_on_near(mid_slot) {
+                Ok(true) => start_slot = mid_slot,
+                Ok(false) => last_slot = mid_slot,
+                Err(_) => {
+                    let (lft_slot, val) = self.find_left_non_error_slot(mid_slot, last_slot);
+                    if val == true {
+                        start_slot = lft_slot;
+                    } else {
+                        last_slot = mid_slot;
+                    }
+                }
             }
         }
 
         Ok(start_slot)
     }
 
     fn get_last_slot(&mut self) -> Result<u64, Box<dyn Error>> {
-        trace!(target: "relay", "= Search for last slot on near =");
-
-        let mut slot = self.eth_client_contract.get_last_submitted_slot();
+        debug!(target: "relay", "= Search for last slot on near =");
 
         let finalized_slot = self.eth_client_contract.get_finalized_beacon_block_slot()?;
         trace!(target: "relay", "Finalized slot on near={}", finalized_slot);
 
-        slot = max(finalized_slot, slot);
+        let last_submitted_slot = self.eth_client_contract.get_last_submitted_slot();
+        trace!(target: "relay", "Last submitted slot={}", last_submitted_slot);
+
+        let mut slot = max(finalized_slot, last_submitted_slot);
         trace!(target: "relay", "Init slot for search as {}", slot);
 
-        if self.block_known_on_near(slot)? {
-            self.search_slot_forward(slot)
+        let last_eth_slot = self.beacon_rpc_client.get_last_slot_number()?.as_u64();
+
+        if slot == finalized_slot || self.block_known_on_near(slot)? {
+            if self.enable_binsearch {
+                return self.search_slot_forward(slot, last_eth_slot + 1);
+            } else {
+                while slot < last_eth_slot {
+                    match self.block_known_on_near(slot + 1) {
+                        Ok(true) => slot += 1,
+                        Ok(false) => break,
+                        Err(_) => slot += 1,
+                    }
+                }
+            }
         } else {
-            self.search_slot_backward(finalized_slot, slot)
+            if self.enable_binsearch {
+                return self.binsearch_slot_range(finalized_slot, slot);
+            } else {
+                while slot > finalized_slot {
+                    match self.block_known_on_near(slot) {
+                        Ok(true) => break,
+                        Ok(false) => slot -= 1,
+                        Err(_) => slot -= 1
+                    }
+                }
+            }
         }
```

Comment:
Move it to the previous `else`, one branch is redundant.
So it will be

```suggestion
        } else if self.enable_binsearch {
            return self.binsearch_slot_range(finalized_slot, slot);
        } else {
            while slot > finalized_slot {
                match self.block_known_on_near(slot) {
                    Ok(true) => break,
                    Ok(false) => slot -= 1,
                    Err(_) => slot -= 1
                }
            }
        }
```

Reply:
Technically yes. But this change does not represent the code logic. What do you think about this variant: https://github.com/aurora-is-near/rainbow-bridge/pull/775/commits/c8b3ab86665059a066619deef25009297d6f08b5

Reply:
It would be the same code logic for that piece of code. However, your new implementation looks better and more refactored.